resource/graph_node.go

```diff
@@ -0,0 +1,309 @@
+package resource
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+	"go.viam.com/utils"
+)
+
+// A GraphNode contains the current state of a resource.
+// It starts out as either uninitialized, unconfigured, or configured.
+// Based on these states, the underlying Resource may or may not be available.
+// Additionally, the node can be informed that the resource either needs to be
+// updated or eventually removed. During its life, errors may be set on the
+// node to indicate that the resource is no longer available to external users.
+type GraphNode struct {
+	mu                        sync.RWMutex
+	updatedAt                 int64
+	clock                     *atomic.Int64
+	current                   Resource
+	currentModel              Model
+	config                    Config
+	needsReconfigure          bool
+	lastErr                   error
+	markedForRemoval          bool
+	unresolvedDependencies    []string
+	needsDependencyResolution bool
+}
+
+var (
+	errNotInitalized  = errors.New("resource not initialized yet")
+	errPendingRemoval = errors.New("resource is pending removal")
+)
+
+// NewUninitializedNode returns a node that is brand new and not yet initialized.
+func NewUninitializedNode() *GraphNode {
+	return &GraphNode{}
+}
+
+// NewUnconfiguredGraphNode returns a node that contains enough information to
+// construct the underlying resource.
+func NewUnconfiguredGraphNode(config Config, dependencies []string) *GraphNode {
+	node := &GraphNode{}
+	node.SetNewConfig(config, dependencies)
+	return node
+}
+
+// NewConfiguredGraphNode returns a node that is already configured with
+// the supplied config and resource.
+func NewConfiguredGraphNode(config Config, res Resource, resModel Model) *GraphNode {
+	node := &GraphNode{}
+	node.SetNewConfig(config, nil)
+	node.setDependenciesResolved()
+	node.SwapResource(res, resModel)
+	return node
+}
+
+// UpdatedAt returns the logical time this node was added at.
+func (w *GraphNode) UpdatedAt() int64 {
+	return w.updatedAt
+}
+
+func (w *GraphNode) setClock(clock *atomic.Int64) {
+	w.clock = clock
+}
+
+// Resource returns the underlying resource if it is not pending removal,
+// has no error on it, and is initialized.
+func (w *GraphNode) Resource() (Resource, error) {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+	if w.markedForRemoval {
+		return nil, errPendingRemoval
+	}
+	if w.lastErr != nil {
+		return nil, w.lastErr
+	}
+	if w.current == nil {
+		return nil, errNotInitalized
+	}
```

Comment:
Doesn't this imply "Not Configured" instead of "Not Initialized"? Feels like there's a case where an Unconfigured state could happen here.

Reply:
it does but it doesn't really matter in terms of functionality. this could be updated to handle for the sake of error messages.

Reply:
Fair enough. I'm looking at it through the lens of a discussion I had with Benji earlier, about module reconfiguration/recovery-from-crash, and there's a lot here that can be useful for that, since the nodes now store the config/deps, the module manager itself may not half to. E.g. crash recovery can maybe reinsantiate all the resources just by marking nodes back to "needsUpdate" in the graph, and so that intermediate state of Unconfigured might become important.

tl;dr no worries for now, but this (or something adjacent) may be something Benji wants to tweak for module recovery.

Reply:
got it. yeah in that case it may be a good addition once we get to that work. I don't think it complicates anything!